platforms/jvm/testing-jvm-infrastructure/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestEventAdapter.java

```diff
@@ -117,9 +119,31 @@ private void reportFailure(Object descriptorId, Throwable throwable) {
 
     @Override
     public void testAssumptionFailure(Failure failure) {
+        TestDescriptorInternal testInternal;
         synchronized (lock) {
+            testInternal = executing.get(failure.getDescription());
             assumptionFailed.add(failure.getDescription());
         }
+
+        if (testInternal != null) {
+            // This is the normal path, we've just seen a test failure
+            // for a test that we saw start
+            Throwable exception = failure.getException();
+            reportAssumptionFailure(testInternal.getId(), exception);
+        } else {
+            // This can happen when, for example, a @BeforeClass or @AfterClass method fails
+            // We generate an artificial start/failure/completed sequence of events
+            testInternal = nullSafeDescriptor(idGenerator.generateId(), failure.getDescription());
+            resultProcessor.started(testInternal, startEvent());
+            Throwable exception = failure.getException();
+            reportAssumptionFailure(testInternal.getId(), exception);
+            resultProcessor.completed(testInternal.getId(), new TestCompleteEvent(clock.getCurrentTime(), TestResult.ResultType.SKIPPED));
```

Comment:
What happens if you call `resultProcessor.failure` and `resultProcessor.completed` on the same test ID?

Reply:
That's exactly what usually happens.  failure isn't a terminal state, it's just a marker that the test had failures. You still have to call completed on it later.